File: Back-tracking/77-combination/go/77.go
package main

import	"fmt"

var	result [][]int
var path []int

func appendCopy(result [][]int, path []int) [][]int {
	tmp := make([]int, len(path))
	copy(tmp, path)
	return append(result, tmp)
}

func backTracking(n int, start int, k int) {
	if len(path) == k {
		result = appendCopy(result, path)
		return
	}

	for i := start; i <= n; i++ {
		path = append(path, i)
		backTracking(n, i + 1, k)
		path = path[:len(path)-1]
	} 
}

func combine(n int, k int) [][]int {
	result = make([][]int, 0)
	path = make([]int, 0)
	backTracking(n, 1, k)

	return result
}

func main() {
	var n int
	var k int
	fmt.Printf("Please input n and k: ")
	fmt.Scanf("%d %d", &n, &k)
	fmt.Println("n, k: ", n, k)
	combine(n, k)
	fmt.Println(result)
}

File: Back-tracking/77-combination/77.py
from typing import List
import copy


class Solution:
    def __init__(self):
        self.res: List[List[int]] = []
        self.path: List[int] = []

    def back_tracking(self, n: int, k: int, start: int):
        if len(self.path) == k:
            self.res.append(self.path.copy())
            return

        for i in range(start, n + 1):
            self.path.append(i)
            self.back_tracking(n, k, i + 1)
            self.path.pop()

    def combine(self, n: int, k: int) -> List[List[int]]:
        self.back_tracking(n, k, 1)
        return self.res
